text/0032-improved-api-tokens.md

```diff
@@ -0,0 +1,163 @@
+  - columns should allow `null` values for now.
+- Remove displaying the plaintext token in the frontend, and only display newly created tokens once.
+- Implement new logic to generate the new token pattern.
+  - The `token_version` should be set as `2`.
+- Implement a backwards compatible means of verifying legacy tokens. This should:
+  - Verify the legacy token is valid/invalid.
+  - If valid, hash the legacy token and store it in the table.
+  - Remove the plaintext representation of the token in the table.
+  - Mark the `token_version` as `1`.
+- A notification/warning in the UI should be displayed recommending users recreate their tokens, resulting in the new token version.
+
+The `n+1` Sentry version should run in production for sufficient time to update the majority of the rows in the database as the legacy tokens are _used_ by users.
+
+### `n+2`
+
+- Add a Django migration to generate the `hashed_token` value for any tokens in `sentry_apitoken` that do not already have a hashed value.
+  - Rows matching this are legacy tokens that were not used during the transition period.
+  - The `token_version` should be set to `1` for these rows.
+
+## Option #2
+
+Instead of slowly generating the hashed token values over time of the legacy tokens (as in Option #1), we could generate a single migration that migrates the entire table to the end state.
+
+## Option #3
+
+To avoid the two different token versions, we could automatically prepend the prefix `sntyx_` (with `x` just being a placeholder here). We would then follow a similar approach to Option #1 or Option #2 to generate the hashed values.
+
+# Drawbacks
+
+## Option #2
+
+- This is less than ideal because of the potential performance impact while the migration mutates each row in the table. Potential impacts to self-hosted installs would be unknown as well.
+
+## Option #3
+
+- Users would not be able to benefit from the Github Secrets Scanning since they would still be using their 64 character alphanumeric string without the prefix.
+- Authentication views/logic would become more complex with branching code to handle the with and without prefix cases.
+
+# Unresolved questions
+
+- How do we deploy the necessary changes for the backend and frontend separately and in a backwards compatible way?
+- Implementation path for `APIApplication` secrets.
+- Is there a downside to allowing `null` in the additional columns?
+- What is the median time between token use?
+  - _This value could be used to inform how long we wait between versions for the migration that will edit pending rows in the database._
+- Do we want to support [API Keys](https://docs.sentry.io/api/auth/#api-keys) as well, even though they are considered a legacy means of authentication?
+- Are there additional secret types we would like to apply this to?
+- Should the secret verification service live within the monolith or run as a separate service entirely?
```

Comment:
If it is going to change the monolith database, we should keep the code in the monolith. Otherwise we end up with an integration database :cry:

Large batches are fine we can process the batch in slices asynchronously with celery.